src/algebra/commute.lean

```diff
@@ -0,0 +1,250 @@
+`(hb.pow 2).add (hb.mul hc)` proves `commute a (b ^ 2 + b * c)`.
+Lean does not immediately recognise these terms as equations,
+so for rewriting we need syntax like `rw [(hb.pow_left 5).eq]`
+rather than just `rw [hb.pow_left 5]`.
+-/
+
+import algebra.group_power
+  group_theory.submonoid group_theory.subgroup
+  ring_theory.subring
+
+section semigroup
+
+variables {S : Type*} [semigroup S]
+
+@[reducible] def commute (a b : S) : Prop := a * b = b * a
+
+theorem commute.eq {a b : S} (h : commute a b) : a * b = b * a := h
+
+@[refl] theorem commute.refl (a : S) : commute a a := by { unfold commute }
+
+@[symm] theorem commute.symm {a b : S} : commute a b → commute b a :=
+λ h, h.symm
+
+theorem commute.mul {a b c : S} :
+ commute a b → commute a c → commute a (b * c) :=
+λ hab hac,
+  by { dsimp [commute] at *,
+       rw [mul_assoc, ← hac, ← mul_assoc b, ← hab, mul_assoc] }
+
+theorem commute.mul_left {a b c : S} :
+ commute a c → commute b c → commute (a * b) c :=
+λ hac hbc, (hac.symm.mul hbc.symm).symm
+
+def centralizer (a : S) : set S := { x | commute a x }
+
+theorem mem_centralizer (a : S) {x : S} : x ∈ centralizer a ↔ commute a x :=
+iff.refl _
+
+def set_centralizer (T : set S) : set S :=
+  { x | ∀ (a : S), a ∈ T → commute a x }
+
+theorem mem_set_centralizer (T : set S) {x : S} :
+  x ∈ set_centralizer T ↔ ∀ (a : S), a ∈ T → commute a x :=
+iff.refl _
+
+end semigroup
+
+section monoid
+
+variables {M : Type*} [monoid M]
+
+theorem commute.one (a : M) : commute a 1 :=
+by { dsimp [commute], rw [one_mul, mul_one] }
+
+theorem commute.one_left (a : M) : commute 1 a := (commute.one a).symm
+
+theorem commute.pow {a b : M} (hab : commute a b) : ∀ (n : ℕ), commute a (b ^ n)
+| 0 := by { rw [pow_zero], exact commute.one a }
+| (n + 1) := by { rw [pow_succ], exact hab.mul (commute.pow n) }
+
+theorem commute.pow_left {a b : M} (hab : commute a b) (n : ℕ) : commute (a ^ n) b :=
+(hab.symm.pow n).symm
+
+theorem commute.pow_pow {a b : M} (hab : commute a b) (n m : ℕ) :
+ commute (a ^ n) (b ^ m) :=
+commute.pow (commute.pow_left hab n) m
+
+theorem commute.self_pow (a : M) (n : ℕ) : commute a (a ^ n) :=
+(commute.refl a).pow n
+
+theorem commute.pow_self (a : M) (n : ℕ) : commute (a ^ n) a :=
+(commute.refl a).pow_left n
+
+theorem commute.pow_pow_self (a : M) (n m : ℕ) : commute (a ^ n) (a ^ m) :=
+(commute.refl a).pow_pow n m
+
+instance centralizer.is_submonoid (a : M) : is_submonoid (centralizer a) :=
+{ one_mem := commute.one a,
+  mul_mem := λ x y hx hy, hx.mul hy }
+
+instance set_centralizer.is_submonoid (S : set M) : is_submonoid (set_centralizer S) :=
+{ one_mem := λ a h, commute.one a,
+  mul_mem := λ x y hx hy a h, (hx a h).mul (hy a h) }
+
+end monoid
+
+section comm_monoid
+
+variables {M : Type*} [comm_monoid M]
+
+theorem commute.all (a b : M) : commute a b := mul_comm a b
+
+end comm_monoid
+
+section group
+
+variables {G : Type*} [group G]
+
+theorem commute.inv {a b : G} (hab : commute a b) : commute a b⁻¹ :=
```

Comment:
This is true for a `[monoid G]`, if you take `a : G`, `b : units G`.

Reply:
Sorry, it implies that `{b : units G | b.val ∈ centralizer a}` is a subgroup of `units G`.

Reply:
I mean something like https://github.com/NeilStrickland/mathlib/pull/1

Reply:
I have put in a version of this.